### ISSUE TITLE
com.hazelcast.client.listeners.leak.ListenerLeakTest.testBackupListenerRemoved_afterClientShutdown

### ISSUE THREAD

**Issue by ahmetmircik** [hazelcast/hazelcast#15755]
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3785/testReport/junit/com.hazelcast.client.listeners.leak/ListenerLeakTest/testBackupListenerRemoved_afterClientShutdown_smartRouting_true_/

```
Stacktrace
java.lang.AssertionError: expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.client.listeners.leak.ListenerLeakTest.testBackupListenerRemoved_afterClientShutdown(ListenerLeakTest.java:268)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by mdogan** [comment]
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3788/testReport/junit/com.hazelcast.client.listeners.leak/ListenerLeakTest/testBackupListenerRemoved_afterClientShutdown_smartRouting_true_/

**Comment by mdogan** [comment]
http://jenkins.hazelcast.com/job/Hazelcast-pr-builder/3792/testReport/junit/com.hazelcast.client.listeners.leak/ListenerLeakTest/testBackupListenerRemoved_afterClientShutdown_smartRouting_true_/

```
java.lang.AssertionError: expected:<0> but was:<1>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:645)
	at org.junit.Assert.assertEquals(Assert.java:631)
	at com.hazelcast.client.listeners.leak.ListenerLeakTest.testBackupListenerRemoved_afterClientShutdown(ListenerLeakTest.java:268)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:114)
	at com.hazelcast.test.FailOnTimeoutStatement$CallableStatement.call(FailOnTimeoutStatement.java:106)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

**Comment by ihsandemir** [comment]
I examined the log for the failure case and I was able to generate the case by putting a delay in method `ClientEndpointImpl.clearAllListeners()`. I put 1 second sleep at the beginning of this method and the test fails with the same reasoning. 

The problem is the test design. It checks for the number of backup listeners at the server side just after client shutdown but the server side endpoint cleanup may not be completed yet. It needs to check for number of backup listener become 0 **eventually**. 

I will submit a PR.
